### DISCUSSION TITLE
Make combined waveform volume adjustable similar to how the filter curve is adjustable

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [libsidplayfp/libsidplayfp#110] [Ideas]
Listening to some tunes that use combined waveforms, the intended volume for these voices differs significantly between composers. There appears to be as much variation between chips as with the filters.

/MUSICIANS/C/Cooksey_Mark/Storm_Warrior.sid, subtune 1 demonstrates this very well.

From 0:00 to 0:41, the first two voices play a tri/pls combo, and the bass on the third voice seems a bit loud.
Then, at 0:41, the first voice uses only the tri and, as a result, becomes much louder, drowning out voice 2 completely.

I've experimented a bit and found that lowering the threshold parameter here:

https://github.com/libsidplayfp/libsidplayfp/blob/206a9d9c8770a9d253c7c9d6878674aa2cc6fd86/src/builders/residfp-builder/residfp/WaveformCalculator.cpp#L258

yields the best results. To test with this particular tune, I've multiplied the value with 0.5 and now the whole songs a lot more balanced and enjoyable.

Another example:

/MUSICIANS/M/Mixer/Dawn.sid

Voice 2 uses combined waveforms for the main melody (saw/pls) and is almost completely inaudible. Multiplying the threshold with 0.8 changes it to a completely different tune.

This is the same problem as with the filters. We can only guess what the original intention was, but it surely wasn't that voices become completely inaudible. There should always be a good balance between melody, bass drums, and other accompanying instruments.

**Comment by Toddler-Boy** [comment]
Addendum: both test-tunes work well with 0.8, so this could be a good default. I will find and test more tunes.

**Comment by drfiemost** [comment]
In latest version I've added a couple of alternative set of parameters, not yet selectable.

https://github.com/libsidplayfp/libsidplayfp/blob/206a9d9c8770a9d253c7c9d6878674aa2cc6fd86/src/builders/residfp-builder/residfp/WaveformCalculator.cpp#L107-L131

I'm not confident enough in the model to start tweaking parameters, there's too much variation.

**Reply by Toddler-Boy** [reply]
> there's too much variation.

And precisely that is the problem. That's why I'm suggesting making these parameters tweakable. Changing the threshold affects the timbre AND volume (less bits make it sound thinner and quieter, more bits fuller and louder).

We will never, ever achieve perfection because it's unknowable. We can never replicate the original author's intent. Ever. In an ideal world, we could reach out to all of them, get their original C64/C128, and measure the heck out of them, but many of these machines have been lost to time, authors have passed away, etc.

It's the filter-offset problem all over again. So, the best we can do is adjust the timbre/volume via a function and provide a "chip-profile" table with a look-up for each author. We must use "best guess" until the majority (or all tunes) from that author sound good.

Using my real-time visualization of the chip internals could be helpful. Suppose you see a voice playing with combined waveforms but don't hear anything (or faintly). In that case, it's a good candidate to add a chip profile for that author, adjust the threshold (or other parameter with a similar effect), and try a few more tunes from the same author until satisfied.

Can you run Windows builds, or do you require a Linux version? I don't have a Linux build yet, but the UI framework I'm using supports it, so building it at some point is just a matter of setting up a virtual machine and adjusting some settings. A windows version I can supply immediately. A macOS version I can also provide within a couple of days.

**Comment by Toddler-Boy** [comment]
Next problem: since most (if not all) modern SIDs are made on VICE with default settings, the combined waveforms in libsidplayfp are now **too** much like real hardware. It leads to 8580 tunes sounding VERY different now.

In this case the added "topbit" is to blame. If I comment out that one line, it sounds like before and sounds like VICE. An alternative is to multiply the threshold by 0.5. Then it sounds 95% like VICE, but more "natural".

/MUSICIANS/M/Mitch_and_Dane/Dane/Uncensored_part_2.sid

The first voice, right at 0:00, plays a very nice bass with a slowly rising cutoff. Listen to it with vsid or sidplayfp 2.5.0. Now compare with sidplayfp 2.6.0. The first voice cuts in and out, because it rapidly alternates between tri/saw/pls (with overlaps, so cw). If I remove the topbit line, it sounds like before. Or if I multiply the threshold parameter with 0.5, I get a similar result, but now you can faintly hear the switching waveforms, which sounds good to me.

**Comment by drfiemost** [comment]
VICE uses samples as is and not a model.

Anyway I'm more into offering a few alternatives, like weak/average/strong CWs, than letting change the parameters.

**Reply by Toddler-Boy** [reply]
So are the current 6581 parameters considered weak, average, or strong? Will there also be weak, average, and strong parameters for the 8580? Although 8580s should be much closer to each other, there still seems to be some kind of audible variation after all.

**Reply by drfiemost** [reply]
Current ones are average.
Can't say for the 8580 as not many samplings are available.

**Reply by Toddler-Boy** [reply]
I'm trying to get my hands on a few more 8580s and measure them.

**Reply by Toddler-Boy** [reply]
Quick update: I've picked 15 different 8580 from 1987 up to 1993 from a seller in the Netherlands on eBay. Hopefully, they can be ordered and delivered soon to get a more representative variation of 8580.

I've tried to pick two from each year, around half a year apart from each other. For 1993, he has only one.

**Comment by Toddler-Boy** [comment]
I just received a C128 today and immediately opened it up. I've pulled the ```6581 R4AR 3586 S``` from it, put it in my PAL C64, heated it up for 30 minutes and then ran the tool again:

![6581R4AR 3586 S](https://github.com/libsidplayfp/libsidplayfp/assets/2496353/32a79efc-1392-4980-9680-b14c4d72eb14)
[6581R4AR 3586 S.zip](https://github.com/libsidplayfp/libsidplayfp/files/14070105/6581R4AR.3586.S.zip)

The 15 different 8580 are still on their way. Shipping is extremely slow :(

**Comment by Toddler-Boy** [comment]
The 15x 8580 arrived today. I've tested them all for functionality and they all work fine. Now I'm measuring each one and upload the results here in one zip. It might take me a day or two, since the process is very tedious and boring.

**Comment by drfiemost** [comment]
That's a nice collection of chips!

**Comment by Toddler-Boy** [comment]
I've measured the first five already. No idea what to do with them once I'm done. I will probably keep a few (one weak, one average, one strong) and give the rest away. Do you want any?

**Comment by Toddler-Boy** [comment]
Here are all 15 chips. Unfortunately, the seller picked a few wrong ones, so I have two ```5092``` (```5092``` & ```5092 #2```). But there are different chips, so who knows?

All are from the same factory, so they are all ```8580 R5 xxxx 25```

[Combined waveforms.zip](https://github.com/libsidplayfp/libsidplayfp/files/14187094/Combined.waveforms.zip)

Please let me know the weakest, the most average, and the strongest one. I would then get rid of the others. If you email me your address, I will send you three for free, and you pick the ones you want.

**Comment by drfiemost** [comment]
As you have seen the strongest one is the 1489 while the weakest is the 4887. There are a few others with a strong topbit but all varies to different degree. You can have a look yourself by using the dump tool at https://github.com/libsidplayfp/combined-waveforms/tree/master/new, it produces a csv file for each cw where you can compare the output values.

And thank you for the offer but I already have a couple of 8580s so I'd say you'd better off selling them on the bay ;)
